### ISSUE TITLE
Strange interaction between RSpec and FFI in 3.2

### ISSUE THREAD

**Issue by ledbettj** [rspec/rspec-mocks#923]
I am using RSpec to stub FFI methods to simulate failing native code ([example](https://github.com/ledbettj/systemd-journal/blob/develop/spec/systemd/journal_spec.rb#L44-L48)) which works as expected in rspec-mocks 3.1.3.

After upgrading to 3.2, FFI gets very upset about the type of something and [throws an ArgumentError](https://github.com/ffi/ffi/blob/master/ext/ffi_c/AbstractMemory.c#L610-L621):

``` 
    Failure/Error: expect { j.disk_usage }.to raise_error(Systemd::JournalError)
       expected Systemd::JournalError, got #<ArgumentError: Invalid Memory object> with backtrace:
         # ./lib/systemd/journal.rb:171:in `disk_usage'
         # ./spec/systemd/journal_spec.rb:48:in `block (4 levels) in <top (required)>'
         # ./spec/systemd/journal_spec.rb:48:in `block (3 levels) in <top (required)>'
```

This happens no matter what when stubbing a FFI function call.

I've poked around at the changes between 3.1.3 and 3.2 as well as at the code under GDB to no avail.  Was there some change here that could be cause this, or some work around?

**Comment by JonRowe** [comment]
Can you provide us with the full backtrace? (add `--backtrace` to your rspec call)

**Comment by ledbettj** [comment]
Whoops!  Yes:

```
  2) Systemd::Journal disk_usage throws a JournalError on invalid return code
     Failure/Error: expect { j.disk_usage }.to raise_error(Systemd::JournalError)
       expected Systemd::JournalError, got #<ArgumentError: Invalid Memory object> with backtrace:
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-support-3.2.2/lib/rspec/support/fuzzy_matcher.rb:15:in `=='
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-support-3.2.2/lib/rspec/support/fuzzy_matcher.rb:15:in `values_match?'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-support-3.2.2/lib/rspec/support/fuzzy_matcher.rb:31:in `block in arrays_match?'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-support-3.2.2/lib/rspec/support/fuzzy_matcher.rb:30:in `each'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-support-3.2.2/lib/rspec/support/fuzzy_matcher.rb:30:in `all?'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-support-3.2.2/lib/rspec/support/fuzzy_matcher.rb:30:in `arrays_match?'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-support-3.2.2/lib/rspec/support/fuzzy_matcher.rb:12:in `values_match?'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-mocks-3.2.1/lib/rspec/mocks/argument_list_matcher.rb:58:in `args_match?'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-mocks-3.2.1/lib/rspec/mocks/message_expectation.rb:398:in `matches?'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-mocks-3.2.1/lib/rspec/mocks/proxy.rb:231:in `block in find_matching_expectation'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-mocks-3.2.1/lib/rspec/mocks/proxy.rb:245:in `block in find_best_matching_expectation_for'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-mocks-3.2.1/lib/rspec/mocks/proxy.rb:244:in `each'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-mocks-3.2.1/lib/rspec/mocks/proxy.rb:244:in `find_best_matching_expectation_for'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-mocks-3.2.1/lib/rspec/mocks/proxy.rb:230:in `find_matching_expectation'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-mocks-3.2.1/lib/rspec/mocks/proxy.rb:159:in `message_received'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-mocks-3.2.1/lib/rspec/mocks/proxy.rb:313:in `message_received'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-mocks-3.2.1/lib/rspec/mocks/method_double.rb:73:in `proxy_method_invoked'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-mocks-3.2.1/lib/rspec/mocks/method_double.rb:60:in `block (2 levels) in define_proxy_method'
         # ./lib/systemd/journal.rb:171:in `disk_usage'
         # ./spec/systemd/journal_spec.rb:48:in `block (4 levels) in <top (required)>'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-expectations-3.2.1/lib/rspec/matchers/built_in/raise_error.rb:43:in `call'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-expectations-3.2.1/lib/rspec/matchers/built_in/raise_error.rb:43:in `matches?'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-expectations-3.2.1/lib/rspec/expectations/handler.rb:50:in `block in handle_matcher'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-expectations-3.2.1/lib/rspec/expectations/handler.rb:27:in `with_matcher'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-expectations-3.2.1/lib/rspec/expectations/handler.rb:48:in `handle_matcher'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-expectations-3.2.1/lib/rspec/expectations/expectation_target.rb:54:in `to'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-expectations-3.2.1/lib/rspec/expectations/expectation_target.rb:87:in `to'
         # ./spec/systemd/journal_spec.rb:48:in `block (3 levels) in <top (required)>'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/example.rb:177:in `instance_exec'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/example.rb:177:in `block in run'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/example.rb:385:in `block in with_around_and_singleton_context_hooks'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/example.rb:343:in `block in with_around_example_hooks'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/hooks.rb:474:in `block in run'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/hooks.rb:612:in `run_around_example_hooks_for'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/hooks.rb:474:in `run'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/example.rb:343:in `with_around_example_hooks'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/example.rb:385:in `with_around_and_singleton_context_hooks'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/example.rb:174:in `run'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/example_group.rb:549:in `block in run_examples'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/example_group.rb:545:in `map'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/example_group.rb:545:in `run_examples'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/example_group.rb:512:in `run'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/example_group.rb:513:in `block in run'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/example_group.rb:513:in `map'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/example_group.rb:513:in `run'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/runner.rb:110:in `block (3 levels) in run_specs'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/runner.rb:110:in `map'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/runner.rb:110:in `block (2 levels) in run_specs'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/configuration.rb:1526:in `with_suite_hooks'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/runner.rb:109:in `block in run_specs'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/reporter.rb:62:in `report'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/runner.rb:108:in `run_specs'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/runner.rb:86:in `run'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/runner.rb:70:in `run'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/runner.rb:38:in `invoke'
         # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/exe/rspec:4:in `<top (required)>'
         # /home/john/.rbenv/versions/2.1.2/bin/rspec:23:in `load'
         # /home/john/.rbenv/versions/2.1.2/bin/rspec:23:in `<main>'
     # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-expectations-3.2.1/lib/rspec/expectations/fail_with.rb:29:in `fail_with'
     # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-expectations-3.2.1/lib/rspec/expectations/handler.rb:40:in `handle_failure'
     # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-expectations-3.2.1/lib/rspec/expectations/handler.rb:50:in `block in handle_matcher'
     # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-expectations-3.2.1/lib/rspec/expectations/handler.rb:27:in `with_matcher'
     # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-expectations-3.2.1/lib/rspec/expectations/handler.rb:48:in `handle_matcher'
     # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-expectations-3.2.1/lib/rspec/expectations/expectation_target.rb:54:in `to'
     # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-expectations-3.2.1/lib/rspec/expectations/expectation_target.rb:87:in `to'
     # ./spec/systemd/journal_spec.rb:48:in `block (3 levels) in <top (required)>'
     # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/example.rb:177:in `instance_exec'
     # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/example.rb:177:in `block in run'
     # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/example.rb:385:in `block in with_around_and_singleton_context_hooks'
     # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/example.rb:343:in `block in with_around_example_hooks'
     # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/hooks.rb:474:in `block in run'
     # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/hooks.rb:612:in `run_around_example_hooks_for'
     # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/hooks.rb:474:in `run'
     # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/example.rb:343:in `with_around_example_hooks'
     # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/example.rb:385:in `with_around_and_singleton_context_hooks'
     # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/example.rb:174:in `run'
     # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/example_group.rb:549:in `block in run_examples'
     # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/example_group.rb:545:in `map'
     # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/example_group.rb:545:in `run_examples'
     # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/example_group.rb:512:in `run'
     # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/example_group.rb:513:in `block in run'
     # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/example_group.rb:513:in `map'
     # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/example_group.rb:513:in `run'
     # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/runner.rb:110:in `block (3 levels) in run_specs'
     # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/runner.rb:110:in `map'
     # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/runner.rb:110:in `block (2 levels) in run_specs'
     # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/configuration.rb:1526:in `with_suite_hooks'
     # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/runner.rb:109:in `block in run_specs'
     # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/reporter.rb:62:in `report'
     # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/runner.rb:108:in `run_specs'
     # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/runner.rb:86:in `run'
     # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/runner.rb:70:in `run'
     # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/lib/rspec/core/runner.rb:38:in `invoke'
     # /home/john/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rspec-core-3.2.3/exe/rspec:4:in `<top (required)>'
     # /home/john/.rbenv/versions/2.1.2/bin/rspec:23:in `load'
     # /home/john/.rbenv/versions/2.1.2/bin/rspec:23:in `<main>'
```

**Comment by myronmarston** [comment]
It looks like one of the FFI objects has an invalid definition of `==` that raises errors when given objects of unexpected type rather than returning `false`.  RSpec expects user objects to implement `==` so that it returns true or false, and `==` should not raise an error.

Still, a number of users have run into issues like this and I realized that in many cases we can avoid calling `==` on user objects to avoid the issue.  I pushed some PRs that I hope will address this:

* #924
* rspec/rspec-support#193

Want to give those PRs a shot?
